database-commons/src/main/java/io/cdap/plugin/db/batch/source/DataDrivenETLDBInputFormat.java

```diff
@@ -130,52 +151,92 @@ public Connection createConnection() {
-    final RecordReader dbRecordReader = super.createDBRecordReader(split, conf);
-    return new RecordReader() {
-      @Override
-      public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
-        dbRecordReader.initialize(split, context);
-      }
+    // This code comes from DataDrivenDBInputFormat.java
+    DBConfiguration dbConf = getDBConf();
+    Class<?> inputClass = dbConf.getInputClass();
+    String dbProductName = getDBProductName();
 
-      @Override
-      public boolean nextKeyValue() throws IOException, InterruptedException {
-        return dbRecordReader.nextKeyValue();
+    LOG.debug("Creating db record reader for db product: " + dbProductName);
+    DBRecordReader delegate;
+    try {
+      // use database product name to determine appropriate record reader.
+      if (dbProductName.startsWith("MYSQL")) {
+        // use MySQL-specific db reader.
+        delegate = new MySQLDataDrivenDBRecordReader(split, inputClass,
+                conf, getConnection(), dbConf, dbConf.getInputConditions(),
+                dbConf.getInputFieldNames(), dbConf.getInputTableName());
+      } else {
+        // Generic reader.
+        delegate = new DataDrivenDBRecordReader(split, inputClass,
+                conf, getConnection(), dbConf, dbConf.getInputConditions(),
+                dbConf.getInputFieldNames(), dbConf.getInputTableName(),
+                dbProductName);
       }
 
-      @Override
-      public Object getCurrentKey() throws IOException, InterruptedException {
-        return dbRecordReader.getCurrentKey();
-      }
+      // Get the user-provided timeout and convert to seconds
+      int timeout = (int) (conf.getLong("timeoutMs", 0) / 1000);
 
-      @Override
-      public Object getCurrentValue() throws IOException, InterruptedException {
-        return dbRecordReader.getCurrentValue();
-      }
+      return new DBRecordReaderWithTimeout(split, inputClass, conf,
+              getConnection(), dbConf, dbConf.getInputConditions(), dbConf.getInputFieldNames(),
+              dbConf.getInputTableName(), delegate, timeout);
+    } catch (SQLException ex) {
+      throw new IOException(ex.getMessage());
+    }
+  }
 
-      @Override
-      public float getProgress() throws IOException, InterruptedException {
-        return dbRecordReader.getProgress();
-      }
+  private class DBRecordReaderWithTimeout extends DBRecordReader {
+    private final DBRecordReader delegate;
+    private final int timeout;
 
-      @Override
-      public void close() throws IOException {
-        dbRecordReader.close();
-        try {
-          DriverManager.deregisterDriver(driverShim);
-        } catch (SQLException e) {
-          throw new IOException(e);
-        }
-      }
-    };
-  }
+    DBRecordReaderWithTimeout(DBInputSplit split, Class inputClass, Configuration conf, Connection conn,
```

Comment:
there also seem to be various things being set in the ConnectionConfigAccessor, we should re-use the same class instead of passing the timeout in through setInput()